src/debug.rs

```diff
@@ -23,6 +26,30 @@ pub trait DebugQueryTable {
     fn keys<C>(&self) -> C
     where
         C: FromIterator<Self::Key>;
+
+    /// Get the (current) set of the entries in the query table.
+    fn entries<C>(&self) -> C
+    where
+        C: FromIterator<TableEntry<Self::Key, Self::Value>>;
+}
+
+/// An entry from a query table, for debugging and inspecting the table state.
+pub struct TableEntry<K, V> {
+    /// key of the query
+    pub key: K,
+    /// value of the query, if it is stored
+    pub value: Option<V>,
```

Comment:
At first I was surprised not to see references here, but given that salsa really encourages your key/value types to be clone-able, makes sense.

Reply:
Yep, this whole setup with `FromIterator` instead of returning an iterator of references is unusual, but that does look like the best solution, given internal mutexes and the fact that this is `debug`.